pytensor/tensor/signal/conv.py

```diff
@@ -0,0 +1,170 @@
+
+    def infer_shape(self, fgraph, node, shapes):
+        in1_shape, in2_shape = shapes
+        n = in1_shape[0]
+        k = in2_shape[0]
+        if self.mode == "full":
+            shape = n + k - 1
+        else:  # mode == "valid":
+            shape = maximum(n, k) - minimum(n, k) + 1
+        return [[shape]]
+
+    def L_op(self, inputs, outputs, output_grads):
+        in1, in2 = inputs
+        [grad] = output_grads
+
+        if self.mode == "full":
+            valid_conv = type(self)(mode="valid")
+            in1_bar = valid_conv(grad, in2[::-1])
+            in2_bar = valid_conv(grad, in1[::-1])
+
+        else:  # mode == "valid":
+            full_conv = type(self)(mode="full")
+            n = in1.shape[0]
+            k = in2.shape[0]
+            kmn = maximum(0, k - n)
+            nkm = maximum(0, n - k)
+            # We need mode="full" if k >= n else "valid" for `in1_bar` (opposite for `in2_bar`), but mode is not symbolic.
+            # Instead, we always use mode="full" and slice the result so it behaves like "valid" for the input that's shorter.
+            in1_bar = full_conv(grad, in2[::-1])
+            in1_bar = in1_bar[kmn : in1_bar.shape[0] - kmn]
+            in2_bar = full_conv(grad, in1[::-1])
+            in2_bar = in2_bar[nkm : in2_bar.shape[0] - nkm]
+
+        return [in1_bar, in2_bar]
+
+
+def convolve1d(
+    in1: "TensorLike",
+    in2: "TensorLike",
+    mode: Literal["full", "valid", "same"] = "full",
+) -> TensorVariable:
+    """Convolve two one-dimensional arrays.
+
+    Convolve in1 and in2, with the output size determined by the mode argument.
+
+    Parameters
+    ----------
+    in1 : (..., N,) tensor_like
+        First input.
+    in2 : (..., M,) tensor_like
+        Second input.
+    mode : {'full', 'valid', 'same'}, optional
+        A string indicating the size of the output:
+        - 'full': The output is the full discrete linear convolution of the inputs, with shape (..., N+M-1,).
+        - 'valid': The output consists only of elements that do not rely on zero-padding, with shape (..., max(N, M) - min(N, M) + 1,).
+        - 'same': The output is the same size as in1, centered with respect to the 'full' output.
+
+    Returns
+    -------
+    out: tensor_variable
+        The discrete linear convolution of in1 with in2.
+
+    """
+    in1 = as_tensor_variable(in1)
+    in2 = as_tensor_variable(in2)
+
+    if mode == "same":
+        # We implement "same" as "valid" with padded `in1`.
+        in1_batch_shape = tuple(in1.shape)[:-1]
+        zeros_left = in2.shape[0] // 2
+        zeros_right = (in2.shape[0] - 1) // 2
+        in1 = join(
+            -1,
+            zeros((*in1_batch_shape, zeros_left), dtype=in2.dtype),
+            in1,
+            zeros((*in1_batch_shape, zeros_right), dtype=in2.dtype),
+        )
+        mode = "valid"
+
+    return cast(TensorVariable, Blockwise(Conv1d(mode=mode))(in1, in2))
+
+
+def convolve(
```

Comment:
I'm not in love with the name and docstring. The docstring says it works on two n-dim arrays, but then raises if they're not 1d? But the name doesn't imply anything about the shapes?

Is this something that is meant to be a generic front-end to more functionality in the future?

Reply:
This is supposed to match the scipy one but right now only supports 1d. It raises a NotImplementedError, not a ValueError. The idea is when we add an arbitrary order conv this function will dispatch to those

Reply:
Want me to remove this until we actually support all the cases the scipy one does?

Reply:
I think that's better. As it stands, it doesn't add anything (except confusion)